packages/react/src/components/Grid/RenderAsProps.ts

```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright IBM Corp. 2023
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+import { PropsWithChildren } from 'react';
+import { ReactAttr } from '../../types/common';
+
+type RenderAsExcludedProps<BaseProps> = 'as' | keyof BaseProps;
+
+type RenderAsDefaultProps<BaseProps> = BaseProps & {
+  as?: undefined;
+} & Omit<ReactAttr<HTMLDivElement>, RenderAsExcludedProps<BaseProps>>;
+
+type RenderAsIntrinsicProps<
+  BaseProps,
+  K extends keyof JSX.IntrinsicElements
+> = BaseProps & {
+  as: K;
+} & Omit<JSX.IntrinsicElements[K], RenderAsExcludedProps<BaseProps>>;
+
+type PropsWithClassName<P> = P & { className?: string };
+type PropsWithClassNameAndChildren<P> = PropsWithChildren<
+  PropsWithClassName<P>
+>;
+
+type RenderAsCustomComponentProps<BaseProps, T> =
+  T extends React.JSXElementConstructor<
+    infer P extends PropsWithClassNameAndChildren<any>
+  >
+    ? BaseProps & {
+        as: React.JSXElementConstructor<P>;
+      } & Omit<P, RenderAsExcludedProps<BaseProps>>
+    : never;
+
+export type RenderAsProps<BaseProps, T> =
+  | RenderAsDefaultProps<BaseProps>
+  | (T extends keyof JSX.IntrinsicElements
+      ? RenderAsIntrinsicProps<BaseProps, T>
+      : RenderAsCustomComponentProps<BaseProps, T>);
```

Comment:
This is a dense file that would benefit from extensive commenting to explain the structure and purpose of each type. 

Also, a helper already exists for `as` props:

https://github.com/carbon-design-system/carbon/blob/fa262d5f0f9b2f95075e20d714ff2535d34b3f3e/packages/react/src/types/common.d.ts#L13-L20

Should that be extended with this? Consider lifting this up out of the Grid component file and into that common file. The `as` prop pattern is used across 21 different components. Searching/find all `as: PropTypes` in the codebase you can see the different ways it's used. Having a cohesive approach for them would be great.

Reply:
@tay1orjones thanks for pointing me at the PolymorphicProps type, I wasn't aware of it's existence. I started with the DefinitelyTyped types and wasn't aware we have such a type already available. In fact, I was also not satisfied what DefinitelyTyped provided and was searching for a method to use ComponentProps type, but couldn't find one. PolymorphicProps are clean and elegant.

I've removed RenderAsProps entirely as it was not necessary anymore and rewritten all components to use PolymorphicProps. Please review once more.